feat[ibmsm]: Secret group name resolution and simpler key interpolation

jgawor opened the pull request:
### Description

* Add support for using secret group names (instead of group IDs) as part of a path. For example, instead of `<path:ibmcloud/arbitrary/secrets/groups/123>` one can use `<path:ibmcloud/arbitrary/secrets/groups/myGroupName>`
* Add support for inserting values of non-arbitrary secrets without using the `jsonPath` modifier. For example, instead of `<path:ibmcloud/imported_cert/secrets/groups/123#my-cert-secret#previous | jsonPath {.certificate}>` one can use `<path:ibmcloud/imported_cert/secrets/groups/123/my-cert-secret/certificate#previous>`

The existing behavior  is unchanged.

### Checklist
Please make sure that your PR fulfills the following requirements:
- [x] Reviewed the guidelines for contributing to this repository
- [x] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] Tests for the changes have been updated
- [ ] Are you adding dependencies? If so, please run `go mod tidy -compat=1.21` to ensure only the minimum is pulled in.
- [x] Docs have been added / updated
- [ ] Optional. My organization is added to USERS.md.

### Type of Change
<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] New tests
- [ ] Build/CI related changes
- [x] Documentation content changes
- [ ] Other (please describe)

### Other information
<!-- Please add any additional information that would help reviewers evaluate your PR -->

jgawor committed:
feat[ibmsm]: Secret group name resolution and simpler key interpolation

Signed-off-by: Jarek Gawor <jgawor@gmail.com>

jkayani commented:
Thanks for the PR! I'll try to review within the week. Initial thoughts:

- Being able to use the group name instead of having to use the ID is a great addition - that way it's consistent with how secret interpolation works, where using the secret name can be used instead of the ID

- Being able to interpolate fields of the secret without using `jsonPath` is also nice. It might be a bit confusing that the IBM Cloud SM backend has 2 different path syntax options but I guess as long as the docs are clear it's OK. 

- On that note, please update the docs to describe the changes made here

Again, thanks for the contribution - code, tests and all. Glad to see this could be implemented without breaking existing behavior as well!

jgawor committed:
add docs

Signed-off-by: Jarek Gawor <jgawor@gmail.com>

jgawor commented:
@jkayani I updated the docs. Please take a look.

jkayani commented on pkg/backends/ibmsecretsmanager.go:
```diff
@@ -536,18 +541,64 @@ func storeSecret(secrets *map[string]interface{}, result map[string]interface{})
 	return nil
 }
 
+func (i *IBMSecretsManager) ResolveGroup(group string) (string, error) {
```
Does this function need to be exported outside the package? If not, let's keep it private like the others here

jgawor replied:
fixed.

jkayani commented on pkg/backends/ibmsecretsmanager.go:
```diff
@@ -536,18 +541,64 @@ func storeSecret(secrets *map[string]interface{}, result map[string]interface{})
+		opts := &ibmsm.ListSecretGroupsOptions{}
+		secretGroupCollection, _, err := i.Client.ListSecretGroups(opts)
+		if err != nil {
+			return "", fmt.Errorf("Could not list security groups: %s", err)
```
```suggestion
                        return "", fmt.Errorf("Could not list secret groups: %s", err)
```

jgawor replied:
fixed.

jkayani commented on pkg/backends/ibmsecretsmanager.go:
```diff
@@ -536,18 +541,64 @@ func storeSecret(secrets *map[string]interface{}, result map[string]interface{})
+	// look up group id for group name
+	groupId := i.secretGroups[group]
+	if groupId == "" {
+		return "", fmt.Errorf("No such security group %s", group)
```
```suggestion
                return "", fmt.Errorf("No such secret group %s", group)
```

jgawor replied:
fixed.

jkayani reviewed (commented):
Only minor feedback here - code, tests, and docs all look great. Once feedback's addressed this can be merged

I don't see any GH issue about this - what motivated this change? Does the alt path syntax you're adding here help in certain automation scenarios, etc? Just curious

Thanks!

jgawor committed:
address comments

Signed-off-by: Jarek Gawor <jgawor@gmail.com>

jgawor commented:
@jkayani I addressed your comments and updated the PR. Thanks! As to the motivation for the PR, it was the overall ease of use. For group names, that's simple. Using the group ids is non-portable from one instance of IBM Secrets Manager to another (e.g. in different regions). Also, trying to decipher the security group id from a path (and mapping it to a group in IBM Secrets Manager) is much harder than a name. As to the simplified path syntax, we actually use the vault plugin with two backends: IBM and AWS. Injecting secrets with the AWS Secrets Manager is simple and does not usually `jsonPath` or other modifiers. With the existing IBM backend, the `jsonPath` modifier is pretty much always needed. So, the idea was to remove that requirement and maybe make moving between the backends a bit simpler.

jkayani reviewed (approved):
LGTM, thanks for your patience!

Paging @werne2j for his sign off
